RavenDB-23642 Support for unbounded time queries in Corax between operation.

maciejaszyk opened the pull request:
### Issue link

https://issues.hibernatingrhinos.com/issue/RavenDB-23642

### Additional description
This PR allows querying unbounded time queries via `between` method (exactly the same way as Lucene does). However, the Corax string unbounded queries do not match null values, since Lucene's `between` matches null only because it has indexed `NULL_VALUE`, so it matches them as a string


### Type of change

- [ ] Bug fix
- [ ] Regression bug fix
- [ ] Optimization
- [x] New feature

### How risky is the change?

- [ ] Low 
- [x] Moderate 
- [ ] High
- [ ] Not relevant

### Backward compatibility

- [x] Non breaking change
- [ ] Ensured. Please explain how has it been implemented?
- [ ] Breaking change
- [ ] Not relevant

### Is it platform specific issue?

- [ ] Yes. Please list the affected platforms.
- [ ] No

### Documentation update

- [ ] This change requires a documentation update. Please mark the issue on YouTrack using `Documentation Required` tag.
- [x] No documentation update is needed 

### Testing by Contributor

- [x] Tests have been added that prove the fix is effective or that the feature works
- [ ] Internal classes added to the test class (e.g. entity or index definition classes) have the lowest possible access modifier (preferable `private`) 
- [ ] It has been verified by manual testing
- [ ] Existing tests verify the correct behavior

### Testing by RavenDB QA team

- [ ] This change requires a special QA testing due to possible performance or resources usage implications (CPU, memory, IO). Please mark the issue on YouTrack using `QA Required` tag.
- [x] No special testing by RavenDB QA team is needed

### Is there any existing behavior change of other features due to this change?

- [ ] Yes. Please list the affected features/subsystems and provide appropriate explanation
- [x] No

### UI work

- [ ] It requires further work in the Studio. Please mark the issue on YouTrack using `Studio Required` tag.
- [x] No UI work is needed

maciejaszyk committed:
RavenDB-23642 Support for unbounded time queries in Corax between operation.

maciejaszyk commented:
test this in debug please

maciejaszyk commented:
test this in debug please

arekpalinski commented:
Please check https://jenkins.hibernatingrhinos.com/job/PR_Tests/14479/testReport/junit/SlowTests.Corax/RavenDB_21573/WhenFieldHasDatesWeStillCanUseAnalyzerToFindOtherValues_options___DatabaseMode___Single___SearchEngineMode___Corax_/

ayende commented on src/Raven.Server/Documents/Indexes/Persistence/Corax/QueryOptimizer/CoraxBooleanItem.cs:
```diff
@@ -49,63 +48,121 @@ private CoraxBooleanItem(IndexSearcher searcher, Index index, FieldMetadata fiel
+                var query = Build(indexSearcher, index, field, leftValue, leftOperator, ref streamingOptimization, out bool isNumerical);
+
+                // For numerical queries we will include the null values for backward compatibility.
+                if (isNumerical == false)
+                    return query;
```
Why aren't we including that for everything?

maciejaszyk replied:
Numerical ranges can be done consistently with the Lucene query results (due to the nature of the `NULL_VALUE` string in Lucene).
However, it would be arbitrary where to put the nulls for string values, since in Lucene our range query could just match `NULL_VALUE` in the requested range. So I would rather suggest that, in terms of string matching, the null should be matched by `or X == null`.

ayende replied:
Wouldn't it be better to state that null is ordered first or last (whatever would be consistent) and that is it?
That means that we have internally consistent behavior for Corax

Lucene won't be with us forever

maciejaszyk replied:
I think we can do it consistently to the numeric ranges behavior.  I pushed the change.

maciejaszyk committed:
RavenDB-23642 Corax:
- Change analyzer to exact in the between method only when at least one string is a time value.
- Between method for textual values: add null values when the right side is unbounded.
